xdsl/dialects/experimental/stencil.py

```diff
@@ -0,0 +1,369 @@
+        if len(attr.data) != self.length:
+            raise VerifyException(
+                f"Expected array of length {self.length}, got {len(attr.data)}."
+            )
+
+
+@irdl_attr_definition
+class FieldType(ParametrizedAttribute, MLIRType):
+    name = "stencil.field"
+
+    shape: ParameterDef[ArrayAttr[IntAttr]]
+
+    @staticmethod
+    def from_shape(shape: list[int] | list[IntAttr]) -> FieldType:
+        # TODO: why do we need all these casts here, can we tell pyright "trust me"
+        if all(isinstance(elm, IntAttr) for elm in shape):
+            shape = cast(list[IntAttr], shape)
+            return FieldType([ArrayAttr.from_list(shape)])
+
+        shape = cast(list[int], shape)
+        return FieldType(
+            [ArrayAttr.from_list([IntAttr.from_int(d) for d in shape])])
+
+
+@irdl_attr_definition
+class TempType(ParametrizedAttribute, MLIRType):
+    name = "stencil.temp"
+
+    shape: ParameterDef[ArrayAttr[IntAttr]]
+
+    @staticmethod
+    def from_shape(
+            shape: ArrayAttr[IntAttr] | list[IntAttr] | list[int]) -> TempType:
+        assert len(shape) > 0
+
+        if isinstance(shape, ArrayAttr):
+            return TempType([shape])
+
+        # cast to list
+        shape = cast(list[IntAttr] | list[int], shape)
+
+        if isinstance(shape[0], IntAttr):
+            # the if above is a sufficient type guard, but pyright does not understand :/
+            return TempType([ArrayAttr.from_list(shape)])  # type: ignore
+        shape = cast(list[int], shape)
+        return TempType(
+            [ArrayAttr.from_list([IntAttr.from_int(d) for d in shape])])
+
+    def __repr__(self):
+        repr: str = "stencil.Temp<["
+        for size in self.shape.data:
+            repr += f"{size.data} "
+        repr += "]>"
+        return repr
+
+
+@irdl_attr_definition
+class ResultType(ParametrizedAttribute, MLIRType):
+    name = "stencil.result"
+    elem: ParameterDef[AnyFloat]
+
+    @staticmethod
+    def from_type(float_t: AnyFloat):
+        return ResultType([float_t])
+
+
+@dataclass
+class ArrayLength(AttrConstraint):
+    length: int = 0
+
+    def verify(self, attr: Attribute) -> None:
+        if not isinstance(attr, ArrayAttr):
+            raise VerifyException(
+                f"Expected {ArrayAttr} attribute, but got {attr.name}.")
+        attr = cast(ArrayAttr[Any], attr)
+        if len(attr.data) != self.length:
+            raise VerifyException(
+                f"Expected array of length {self.length}, got {len(attr.data)}."
+            )
+
+
+# TODO: How can we inherit from MLIRType and ParametrizedAttribute?
+@dataclass(frozen=True)
+class Stencil_Element(ParametrizedAttribute):
+    name = "stencil.element"
+    element = AnyOf([f32, f64])
+
+
+@dataclass(frozen=True)
+class Stencil_Index(ParametrizedAttribute):
+    # TODO: can you have an attr and an op with the same name?
+    name = "stencil.index"
+    shape = Annotated[ArrayAttr[IntAttr], ArrayLength(2)]
```

Comment:
Why do you sometimes use this form and sometimes the `ParameterDef` form? We usually only use the later for ParametrizedAttributes, right?

Reply:
Oh, I didn't even notice that we use `ParameterDef` elsewhere!